src/coprocessor/dag/storage/scanner.rs

```diff
@@ -0,0 +1,662 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+use super::range::*;
+use super::ranges_iter::*;
+use super::{OwnedKvPair, Storage};
+use crate::coprocessor::Error;
+
+const KEY_BUFFER_CAPACITY: usize = 64;
+
+/// A scanner that scans over multiple ranges. Each range can be a point range containing only
+/// one row, or an interval range containing multiple rows.
+pub struct RangesScanner<T> {
+    storage: T,
+    ranges_iter: RangesIterator,
+
+    scan_backward_in_range: bool,
+    is_key_only: bool,
+
+    scanned_rows_per_range: Vec<usize>,
+
+    // The following fields are only used for calculating scanned range. Scanned range is only
+    // useful in streaming mode, where the client need to know the underlying physical data range
+    // of each response slice, so that partial retry can be non-overlapping.
+    is_scanned_range_aware: bool,
+    current_range_lower_key: Vec<u8>,
```

Comment:
I am curious why don't use `Range` instead of two individual fields to represent it (any consideration?).

Reply:
Good catch